policy/parser.go

```diff
@@ -0,0 +1,615 @@
+type Policy struct {
+	name     ValueString
+	rule     *Rule
+	semantic semanticType
+	info     *ast.SourceInfo
+	source   *Source
+}
+
+// Source returns the policy file contents as a CEL source object.
+func (p *Policy) Source() *Source {
+	return p.source
+}
+
+// SourceInfo returns the policy file metadata about expression positions.
+func (p *Policy) SourceInfo() *ast.SourceInfo {
+	return p.info
+}
+
+// Name returns the name of the policy.
+func (p *Policy) Name() ValueString {
+	return p.name
+}
+
+// Rule returns the rule entry point of the policy.
+func (p *Policy) Rule() *Rule {
+	return p.rule
+}
+
+// SetName configures the policy name.
+func (p *Policy) SetName(name ValueString) {
+	p.name = name
+}
+
+// SetRule configures the policy rule entry point.
+func (p *Policy) SetRule(r *Rule) {
+	p.rule = r
+}
+
+// Rule declares a rule identifier, description, along with a set of variables and match statements.
+type Rule struct {
+	id          *ValueString
+	description *ValueString
+	variables   []*Variable
+	matches     []*Match
+}
+
+// ID returns the id value of the rule if it is set.
+func (r *Rule) ID() ValueString {
+	if r.id != nil {
+		return *r.id
+	}
+	return ValueString{}
+}
+
+// Description returns the rule description if it is set.
+func (r *Rule) Description() ValueString {
+	if r.description != nil {
+		return *r.description
+	}
+	return ValueString{}
+}
+
+// Matches returns the ordered set of Match declarations.
+func (r *Rule) Matches() []*Match {
+	return r.matches[:]
+}
+
+// Variables returns the order set of Variable tuples.
+func (r *Rule) Variables() []*Variable {
+	return r.variables[:]
+}
+
+// SetID configures the id for the rule.
+func (r *Rule) SetID(id ValueString) {
+	r.id = &id
+}
+
+// SetDescription configures the description for the rule.
+func (r *Rule) SetDescription(desc ValueString) {
+	r.description = &desc
+}
+
+// AddMatch addes a Match to the rule.
+func (r *Rule) AddMatch(m *Match) {
+	r.matches = append(r.matches, m)
+}
+
+// AddVariable adds a variable to the rule.
+func (r *Rule) AddVariable(v *Variable) {
+	r.variables = append(r.variables, v)
+}
+
+// Variable is a named expression which may be referenced in subsequent expressions.
+type Variable struct {
+	name       ValueString
+	expression ValueString
+}
+
+// Name returns the variable name.
+func (v *Variable) Name() ValueString {
+	return v.name
+}
+
+// Expression returns the variable expression.
+func (v *Variable) Expression() ValueString {
+	return v.expression
+}
+
+// SetName sets the variable name.
+func (v *Variable) SetName(name ValueString) {
+	v.name = name
+}
+
+// SetExpression sets the variable expression.
+func (v *Variable) SetExpression(e ValueString) {
+	v.expression = e
+}
+
+// Match declares a condition (defaults to true) as well as an output or a rule.
+// Either the output or the rule field may be set, but not both.
+type Match struct {
+	condition ValueString
```

Comment:
Any reason for condition being a non-pointer?

Reply:
Mostly just that it's required. The others are optional.